### PR TITLE
D3ASIM-3356: Add attributes and requirements arguments to  offer_energy and bid_ennergy commands

### PR DESCRIPTION
**Pull request by Hassan754**

### CODE REVIEW

File: README.md
Comment (fievelk) [inline README.md:301]: Looking at the rendered HTML version, I think that the syntax could be changed for clarity. I'd also remove the types since the `README.md` is not supposed to be an exhaustive source of documentation (I think). For example:
```suggestion
- Offers:
    - `energy_type`: Energy type of the offer
- Bids: No supported bids' attributes at the moment

Currently supported requirements:
- Offers:
    - `preferred_trading_partner`: preferable area ids to match with.
- Bids:
    - `energy_type`: Energy types that the bid prefers to consume.
    - `preferred_trading_partner`: preferable area ids to match with.
```

File: d3a_api_client/commands.py
Comment (fievelk) [inline d3a_api_client/commands.py:29]:
```suggestion
            attributes: Dict = None, requirements: List[Dict] = None):
```
The same applies to the rest of the PR.

**Review by fievelk — COMMENTED** [review]
Two comments.

On a side note: I have the feeling that the bids/offers methods are beginning to receive too many arguments. Sooner or later we might want to refactor this somehow, but it's not necessary now.

**Review by hannesdiedrich — DISMISSED** [review]
Please dont forget to revert the d3a branch in build_test_containers.sh,
LGTM

**Review by spyrostz — DISMISSED** [review]
LGTM, however since this PR deals with it, I would like to reiterate/double check that offers should not have requirements, at least judging from our last discussions with @candrews42 .

**Comment by Hassan754** [comment]
> LGTM, however since this PR deals with it, I would like to reiterate/double check that offers should not have requirements, at least judging from our last discussions with @candrews42 .

Actually I recall that Colin mentioned that them both should have requirements
https://gridsingularity.atlassian.net/browse/D3ASIM-3356?focusedCommentId=22832

**Comment by spyrostz** [comment]
> > LGTM, however since this PR deals with it, I would like to reiterate/double check that offers should not have requirements, at least judging from our last discussions with @candrews42 .
> 
> Actually I recall that Colin mentioned that them both should have requirements
> https://gridsingularity.atlassian.net/browse/D3ASIM-3356?focusedCommentId=22832

It should not, but this part of the comment clarified my question: `allowing preferred trading partner as a requirement`
Therefore preferred trading partner is considered a requirement, and this is the only requirement allowed in the offers.  So all good here (even though it would be good for these restrictions to also be implemented on the client side, can be postponed for later until we have a basic operation though)

**Comment by fievelk** [comment]
> > > LGTM, however since this PR deals with it, I would like to reiterate/double check that offers should not have requirements, at least judging from our last discussions with @candrews42 .
> > 
> > 
> > Actually I recall that Colin mentioned that them both should have requirements
> > https://gridsingularity.atlassian.net/browse/D3ASIM-3356?focusedCommentId=22832
> 
> It should not, but this part of the comment clarified my question: `allowing preferred trading partner as a requirement`
> Therefore preferred trading partner is considered a requirement, and this is the only requirement allowed in the offers (obviously Colin has confused bids with offers in his comment here, because the bids should obviously be able to select e.g. the source of energy they want to consume). So all good here (even though it would be good for these restrictions to also be implemented on the client side, can be postponed for later until we have a basic operation)

Just jumping in on this: if we want to validate requirements/attributes, I suggest we implement some class that can be shared between d3a-api-client and d3a (to avoid having to change both implementations everytime something changes).

**Comment by spyrostz** [comment]
> > > > LGTM, however since this PR deals with it, I would like to reiterate/double check that offers should not have requirements, at least judging from our last discussions with @candrews42 .
> > > 
> > > 
> > > Actually I recall that Colin mentioned that them both should have requirements
> > > https://gridsingularity.atlassian.net/browse/D3ASIM-3356?focusedCommentId=22832
> > 
> > 
> > It should not, but this part of the comment clarified my question: `allowing preferred trading partner as a requirement`
> > Therefore preferred trading partner is considered a requirement, and this is the only requirement allowed in the offers (obviously Colin has confused bids with offers in his comment here, because the bids should obviously be able to select e.g. the source of energy they want to consume). So all good here (even though it would be good for these restrictions to also be implemented on the client side, can be postponed for later until we have a basic operation)
> 
> Just jumping in on this: if we want to validate requirements/attributes, I suggest we implement some class that can be shared between d3a-api-client and d3a (to avoid having to change both implementations everytime something changes).

I wholeheartedly agree on this, even if the validation takes place in 2 places (d3a-api-client and d3a) it definitely makes sense to have it in d3a-interface and reuse it in both places.

**Comment by Hassan754** [comment]
> Two comments.
> 
> On a side note: I have the feeling that the bids/offers methods are beginning to receive too many arguments. Sooner or later we might want to refactor this somehow, but it's not necessary now.

@fievelk I agree, even the next test (Test 3 ) adds a new offer and bid argument (time_slot)
I don't currently have an easy way to fix that, I'll probably push this until later on In the context of test 3 if it won't take lot of time or I would create a dedicated backend ticket for this, your call 😄

**Comment by fievelk** [comment]
> > Two comments.
> > On a side note: I have the feeling that the bids/offers methods are beginning to receive too many arguments. Sooner or later we might want to refactor this somehow, but it's not necessary now.
> 
> @fievelk I agree, even the next test (Test 3 ) adds a new offer and bid argument (time_slot)
> I don't currently have an easy way to fix that, I'll probably push this until later on In the context of test 3 if it won't take lot of time or I would create a dedicated backend ticket for this, your call 😄

I think that the dedicated backend ticket would never make it to a sprint. I would try to tackle this in test 3 if possible. Otherwise the ticket is the only option :)

**Comment by Hassan754** [comment]
> > > Two comments.
> > > On a side note: I have the feeling that the bids/offers methods are beginning to receive too many arguments. Sooner or later we might want to refactor this somehow, but it's not necessary now.
> > 
> > 
> > @fievelk I agree, even the next test (Test 3 ) adds a new offer and bid argument (time_slot)
> > I don't currently have an easy way to fix that, I'll probably push this until later on In the context of test 3 if it won't take lot of time or I would create a dedicated backend ticket for this, your call 😄
> 
> I think that the dedicated backend ticket would never make it to a sprint. I would try to tackle this in test 3 if possible. Otherwise the ticket is the only option :)

If we're given capacity of time in the Test 3, I will handle it, otherwise we have no option besides creating a new ticket for this and give it priority.
This feature imo isn't straightforward, we'll need to handle the fields of class Offer and Bid in the d3a, + the interfaces for posting these offers and bids

**Review by hannesdiedrich — APPROVED** [review]
LGTM

File: integration_tests/build_test_containers.sh
Comment (hannesdiedrich) [inline integration_tests/build_test_containers.sh:17]: please remove before merging

**Review by hannesdiedrich — APPROVED** [review]
LGTM

**Review by faizan2590 — DISMISSED** [review]
LGTM

**Review by spyrostz — APPROVED** [review]
LGTM

**Review by hannesdiedrich — DISMISSED** [review]
LGTM

File: README.md
Comment (fievelk) [inline README.md:294]: Nit: I tried to rephrase this a bit:
```suggestion
### Attributes and requirements

A finer control over the issued bids and offers can be achieved through the `attributes` and `requirements` parameters of the bid and offer methods.

#### `attributes`

A dictionary of attributes that describe the offer or bid. Currently supported attributes:

- Offers:
    - `energy_type`: energy type of the offer

- Bids: Not supported at the moment

#### `requirements`

A list of dictionaries containing requirements for the offer or bid. At least one of the specified requirements needs to be satisfied in the matching process. Currently supported requirements:

- Offers:
    - `trading_partners`: IDs of the areas with which the offer has to be matched

- Bids:
    - `energy_type`: energy types that the bid prefers to consume
    - `trading_partners`: IDs of the areas with which the bid prefers to be matched
    - `energy`: energy that the bid prefers to consume
    - `price`: trade rate that the bid prefers to accept
```
Feel free to use it or discard it if you don't like it!

**Review by fievelk — APPROVED** [review]

**Review by spyrostz — APPROVED** [review]
LGTM

**Review by hannesdiedrich — APPROVED** [review]
LGTM

**Pull request merged**
